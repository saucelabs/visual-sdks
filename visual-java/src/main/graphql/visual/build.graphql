query build($input: UUID!) {
  result: build(id: $input) {
    url
  }
}
