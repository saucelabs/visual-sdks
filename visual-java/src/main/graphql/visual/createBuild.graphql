mutation createBuild($input: BuildIn!) {
  result: createBuild(input: $input) {
    id
    name
    project
    branch
    status
    url
  }
}
