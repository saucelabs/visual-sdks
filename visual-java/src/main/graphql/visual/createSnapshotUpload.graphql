mutation createSnapshotUpload($input: SnapshotUploadIn!) {
  result: createSnapshotUpload(input: $input) {
    id
    buildId
    imageUploadUrl
  }
}
