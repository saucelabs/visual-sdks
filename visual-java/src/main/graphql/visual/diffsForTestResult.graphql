query diffsForTestResult($input: UUID!) {
  result: diffs(condition: { buildId: $input }) {
    nodes {
      id
      status
    }
  }
}
