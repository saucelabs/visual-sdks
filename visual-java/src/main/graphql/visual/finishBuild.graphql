mutation FinishBuild($input: FinishBuildIn!) {
  result: finishBuild(input: $input) {
    id
    name
    status
    url
  }
}
