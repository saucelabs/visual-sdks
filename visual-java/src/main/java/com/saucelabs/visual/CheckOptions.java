package com.saucelabs.visual;

import com.saucelabs.visual.graphql.type.DiffingOptionsIn;
import com.saucelabs.visual.model.DiffingFlag;
import com.saucelabs.visual.model.FullPageScreenshotConfig;
import com.saucelabs.visual.model.IgnoreRegion;
import com.saucelabs.visual.model.VisualRegion;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;
import org.openqa.selenium.WebElement;

public class CheckOptions {

  public enum DiffingMethod {
    SIMPLE,
    EXPERIMENTAL,
    BALANCED,
  }

  public CheckOptions() {}

  public CheckOptions(
      List<WebElement> ignoreElements,
      List<IgnoreRegion> ignoreRegions,
      List<VisualRegion> regions,
      String testName,
      String suiteName,
      DiffingMethod diffingMethod,
      DiffingOptionsIn diffingOptions,
      Boolean captureDom,
      String clipSelector,
      FullPageScreenshotConfig fullPageScreenshotConfig) {
    this.ignoreElements = ignoreElements;
    this.ignoreRegions = ignoreRegions;
    this.regions = regions;
    this.testName = testName;
    this.suiteName = suiteName;
    this.diffingMethod = diffingMethod;
    this.captureDom = captureDom;
    this.clipSelector = clipSelector;
    this.fullPageScreenshotConfig = fullPageScreenshotConfig;
    this.diffingOptions = diffingOptions;
  }

  private List<WebElement> ignoreElements = new ArrayList<>();
  private List<IgnoreRegion> ignoreRegions = new ArrayList<>();
  private List<VisualRegion> regions = new ArrayList<>();

  private String testName;
  private String suiteName;
  private DiffingMethod diffingMethod;
  private DiffingOptionsIn diffingOptions;
  private Boolean captureDom;
  private String clipSelector;

  private FullPageScreenshotConfig fullPageScreenshotConfig;

  public static class Builder {
    private List<WebElement> ignoreElements = new ArrayList<>();
    private List<IgnoreRegion> ignoreRegions = new ArrayList<>();
    private List<VisualRegion> regions = new ArrayList<>();
    private String testName;
    private String suiteName;
    private DiffingMethod diffingMethod;
    private DiffingOptionsIn diffingOptions;
    private Boolean captureDom;
    private String clipSelector;
    private FullPageScreenshotConfig fullPageScreenshotConfig;

    public Builder withIgnoreElements(List<WebElement> ignoreElements) {
      this.ignoreElements = ignoreElements;
      return this;
    }

    public Builder withIgnoreRegions(List<IgnoreRegion> ignoreRegions) {
      this.ignoreRegions = ignoreRegions;
      return this;
    }

    public Builder withTestName(String testName) {
      this.testName = testName;
      return this;
    }

    public Builder withSuiteName(String suiteName) {
      this.suiteName = suiteName;
      return this;
    }

    public Builder withDiffingMethod(DiffingMethod diffingMethod) {
      this.diffingMethod = diffingMethod;
      return this;
    }

    public Builder withCaptureDom(Boolean captureDom) {
      this.captureDom = captureDom;
      return this;
    }

    public Builder withClipSelector(String clipSelector) {
      this.clipSelector = clipSelector;
      return this;
    }

    public Builder withFullPageConfig(FullPageScreenshotConfig fullPageScreenshotConfig) {
      this.fullPageScreenshotConfig = fullPageScreenshotConfig;
      return this;
    }

    public Builder disableOnly(EnumSet<DiffingFlag> flags) {
      this.diffingOptions = new DiffingOptionsIn();
      DiffingFlag.setAll(this.diffingOptions, true);
      for (DiffingFlag f : flags) f.apply(this.diffingOptions, false);
      return this;
    }

    public Builder enableOnly(EnumSet<DiffingFlag> flags) {
      this.diffingOptions = new DiffingOptionsIn();
      DiffingFlag.setAll(this.diffingOptions, false);
      for (DiffingFlag f : flags) f.apply(this.diffingOptions, true);
      return this;
    }

    public Builder enableOnly(EnumSet<DiffingFlag> flags, WebElement element) {

      this.regions.add(VisualRegion.ignoreChangesFor(element).except(flags));
      return this;
    }

    public Builder disableOnly(EnumSet<DiffingFlag> flags, WebElement element) {

      this.regions.add(VisualRegion.detectChangesFor(element).except(flags));
      return this;
    }

    public CheckOptions build() {
      return new CheckOptions(
          ignoreElements,
          ignoreRegions,
          regions,
          testName,
          suiteName,
          diffingMethod,
          diffingOptions,
          captureDom,
          clipSelector,
          fullPageScreenshotConfig);
    }
  }

  public List<WebElement> getIgnoreElements() {
    return ignoreElements;
  }

  public void setIgnoreElements(List<WebElement> ignoreElements) {
    this.ignoreElements = ignoreElements;
  }

  public List<IgnoreRegion> getIgnoreRegions() {
    return ignoreRegions;
  }

  public void setIgnoreRegions(List<IgnoreRegion> ignoreRegions) {
    this.ignoreRegions = ignoreRegions;
  }

  public String getTestName() {
    return testName;
  }

  public void setTestName(String testName) {
    this.testName = testName;
  }

  public List<VisualRegion> getRegions() {
    return regions;
  }

  public void setRegions(List<VisualRegion> regions) {
    this.regions = regions;
  }

  public String getSuiteName() {
    return suiteName;
  }

  public void setSuiteName(String suiteName) {
    this.suiteName = suiteName;
  }

  public void setDiffingMethod(DiffingMethod diffingMethod) {
    this.diffingMethod = diffingMethod;
  }

  public DiffingMethod getDiffingMethod() {
    return diffingMethod;
  }

  public DiffingOptionsIn getDiffingOptions() {
    return diffingOptions;
  }

  public void setCaptureDom(Boolean captureDom) {
    this.captureDom = captureDom;
  }

  public Boolean getCaptureDom() {
    return captureDom;
  }

  public String getClipSelector() {
    return clipSelector;
  }

  public void setClipSelector(String clipSelector) {
    this.clipSelector = clipSelector;
  }

  public FullPageScreenshotConfig getFullPageScreenshotConfig() {
    return fullPageScreenshotConfig;
  }

  public void enableFullPageScreenshots(FullPageScreenshotConfig fullPageScreenshotConfig) {
    this.fullPageScreenshotConfig = fullPageScreenshotConfig;
  }

  public void enableFullPageScreenshots() {
    this.fullPageScreenshotConfig = new FullPageScreenshotConfig.Builder().build();
  }
}
