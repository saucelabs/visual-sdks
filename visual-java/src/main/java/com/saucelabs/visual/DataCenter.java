package com.saucelabs.visual;

public enum DataCenter {
  US_WEST_1("https://api.us-west-1.saucelabs.com/v1/visual/graphql"),
  US_EAST_4("https://api.us-east-4.saucelabs.com/v1/visual/graphql"),
  EU_CENTRAL_1("https://api.eu-central-1.saucelabs.com/v1/visual/graphql");

  public final String endpoint;

  DataCenter(String endpoint) {
    this.endpoint = endpoint;
  }
}
