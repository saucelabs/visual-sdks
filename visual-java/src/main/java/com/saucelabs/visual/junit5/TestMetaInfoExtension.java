package com.saucelabs.visual.junit5;

import com.saucelabs.visual.TestMetaInfo;
import java.util.Optional;
import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
import org.junit.jupiter.api.extension.BeforeTestExecutionCallback;
import org.junit.jupiter.api.extension.ExtensionContext;

public class TestMetaInfoExtension
    implements BeforeTestExecutionCallback, AfterTestExecutionCallback {
  @Override
  public void beforeTestExecution(ExtensionContext context) throws Exception {
    String className = context.getRequiredTestClass().getName();
    String testName = context.getRequiredTestMethod().getName();

    TestMetaInfo.THREAD_LOCAL.set(Optional.of(new TestMetaInfo(className, testName)));
  }

  @Override
  public void afterTestExecution(ExtensionContext context) throws Exception {
    TestMetaInfo.THREAD_LOCAL.set(Optional.empty());
  }
}
