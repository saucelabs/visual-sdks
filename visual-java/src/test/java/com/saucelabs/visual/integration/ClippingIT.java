package com.saucelabs.visual.integration;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.saucelabs.visual.CheckOptions;
import com.saucelabs.visual.junit5.TestMetaInfoExtension;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.openqa.selenium.By;

@ExtendWith({TestMetaInfoExtension.class, SnapshotExtension.class})
public class ClippingIT extends IntegrationBase {
  Expect expect;

  @BeforeAll
  public static void login() {
    LoginPage loginPage = new LoginPage();
    loginPage.open();
    loginPage.login();
    InventoryLongPage inventoryPage = new InventoryLongPage();
    inventoryPage.open();
  }

  @Test
  public void clipToSelectorOnThePage() {
    String id =
        sauceVisualCheck(
            "Clipped selector",
            new CheckOptions.Builder().withClipSelector(".inventory_list").build());
    CheckSnapshotOperation.Result result = getSnapshotResult(id);
    expect.toMatchSnapshot(result.data);
  }

  @Test
  public void clipToElementOnThePage() {
    String id =
        sauceVisualCheck(
            "Clipped element",
            new CheckOptions.Builder()
                .withClipElement(driver.findElement(By.cssSelector(".inventory_list")))
                .build());
    CheckSnapshotOperation.Result result = getSnapshotResult(id);
    expect.toMatchSnapshot(result.data);
  }

  @Test
  public void scrollToElementBeforeClipping() {
    String id =
        sauceVisualCheck(
            "Scroll to element before clipping",
            new CheckOptions.Builder()
                .withClipElement(
                    driver.findElement(By.cssSelector(".inventory_item:nth-child(21)")))
                .build());
    CheckSnapshotOperation.Result result = getSnapshotResult(id);
    expect.toMatchSnapshot(result.data);
  }
}
