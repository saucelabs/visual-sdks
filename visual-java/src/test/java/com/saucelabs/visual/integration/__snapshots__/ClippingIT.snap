com.saucelabs.visual.integration.ClippingIT.clipToElementOnThePage=[
{
  "snapshot" : {
    "name" : "Clipped element (Sauce)",
    "suiteName" : "com.saucelabs.visual.integration.ClippingIT",
    "testName" : "clipToElementOnThePage",
    "operatingSystem" : "WINDOWS",
    "operatingSystemVersion" : "10",
    "device" : "Desktop (1600x1059)",
    "devicePixelRatio" : 1,
    "hasDom" : true,
    "diffs" : {
      "nodes" : [ {
        "status" : "EQUAL",
        "diffClusters" : [ ]
      } ]
    },
    "ignoreRegions" : [ ]
  }
}
]


com.saucelabs.visual.integration.ClippingIT.clipToSelectorOnThePage=[
{
  "snapshot" : {
    "name" : "Clipped selector (Sauce)",
    "suiteName" : "com.saucelabs.visual.integration.ClippingIT",
    "testName" : "clipToSelectorOnThePage",
    "operatingSystem" : "WINDOWS",
    "operatingSystemVersion" : "10",
    "device" : "Desktop (1600x1059)",
    "devicePixelRatio" : 1,
    "hasDom" : true,
    "diffs" : {
      "nodes" : [ {
        "status" : "EQUAL",
        "diffClusters" : [ ]
      } ]
    },
    "ignoreRegions" : [ ]
  }
}
]


com.saucelabs.visual.integration.ClippingIT.scrollToElementBeforeClipping=[
{
  "snapshot" : {
    "name" : "Scroll to element before clipping (Sauce)",
    "suiteName" : "com.saucelabs.visual.integration.ClippingIT",
    "testName" : "scrollToElementBeforeClipping",
    "operatingSystem" : "WINDOWS",
    "operatingSystemVersion" : "10",
    "device" : "Desktop (1600x1059)",
    "devicePixelRatio" : 1,
    "hasDom" : true,
    "diffs" : {
      "nodes" : [ {
        "status" : "EQUAL",
        "diffClusters" : [ ]
      } ]
    },
    "ignoreRegions" : [ ]
  }
}
]